### ISSUE TITLE
HasZeroOrOne Relation

### ISSUE THREAD

**Issue by HolgerW1** [laravel/framework#8376]
It would be great if a HasZeroOrOne relation class would exist, i.e. a class for a 1..[0,1] relation.

Example:
- a user may have a picture (0 or 1), a picture is a file
- (a user may also have 0 or many other files through other relations)

So we've got a user with the fields [id, picture_file_id] and files with [id, user_id]. "picture_file_id" may be null and is the foreign key to the "id" of a file. "user_id" is the foreign key of "files" to the user owning the file.
I.e. what we want to do is associate a file with a user through the "picture_file_id" which is currently not possible with any relation class in laravel (or am I missing sth.?). The relation would have to do two things: set the "picture_file_id" on the parent and set the "user_id" on the child during association.

**Comment by HolgerW1** [comment]
Digging into the HasOne code and why it can't handle the described case the actual problem why HasOne can't handle this case is the following scenario:
user.picture_file_id references file.id
file.user_id references user.id

I.e. if the foreign key of the parent table isn't the same as the referenced key by the foreign table.
So I guess a better solution would be to improve HasOne to handle this case?

**Comment by GrahamCampbell** [comment]
Zero or one won't work because you can't index a nullable column.

**Comment by HolgerW1** [comment]
That's actually not true. MySQL e.g. supports nullable indexed columns unless you use the BDB storage engine (as desribed here https://dev.mysql.com/doc/refman/5.0/en/create-index.html)

**Comment by jarektkaczyk** [comment]
Can you show a code example of what you mean, since I have no idea what cannot be done with currently available relations and what `HasZeroOrOne` is?

In your case `User belongsTo Picture` (no `hasOne`), and if the FK is null, the `picture` relation will be simply `null` - no problem with that at all.

**Comment by HolgerW1** [comment]
tl;dr @jarektkaczyk you're absolutely right you can use `belongsTo` for a 1..[0,1] relation! Thanks for the hint.

But imho for this case it's somehow cumbersome and unintuitive having to use `belongsTo` and having to call
```
$user->pictureFile()->associate($file);
$user->save();
```
since the documentation says `belongsTo` is used for the inverse of a relation.

My suggestion would be to add this case to the documentation so people don't get confused like I did. If you model your architecture the more natural thing would be to define it like "User (1) ---> (0..1) File" (User has zero or one File) instead of "User belongs to (Picture) File".
Plus a `HasOne` relation doesn't need a `save` call on the `User` model, it would be more natural and readable if you could do sth. similar as follows for a `HasZeroOfOne` relation:
```
$user->pictureFile()->save($file);
```

But that's just my opinion, maybe no one else stumbled upon this.

**Comment by jarektkaczyk** [comment]
@HolgerW1 2 things to consider:

1. is it possible that one `File` will be used by many users as `pictureFile`? If so, then that's how the relations should be defined. Thing that matters is where is the field that references the other one, that's all. 

2. If that's not the case, then you should really use `hasOne`:
    ```
    // User
    pictureFile: User hasOne Picture // user has one (or zero) picture file
    pictures: User hasMany Picture // user owns many (or zero) files

    // Picture
    owner: Picture belongsTo User
    ```

**Comment by HolgerW1** [comment]
@jarektkaczyk thanks again for the input.

Here's how my models / relation currently look like:
- A `User` may only have 0 or 1 `File` through the `pictureFile` relation.
```
    public function pictureFile()
    {
        return $this->belongsTo('File', 'picture_file_id');
    }
```
- A `User` may be owner of many other files and may have other relations to `File` objects
```
    public function files()
    {
        return $this->hasMany('File', 'user_id');
    }
```

- A `File` always has exactly one owner, i.e. belongs to a user:
```
    public function user()
    {
        return $this->belongsTo('User');
    }
```

Now as you say the "natural" thing would be to use a `User hasOne File` for the `pictureFile` relation on the `User`:
```
    public function pictureFile()
    {
        return $this->hasOne('File', 'user_id', 'picture_file_id');
    }
```

But if I use that relation and try to save a file through that relation, I do get an integrity constraint violation because the `user_id` of the file is not set properly: 
```
$user->pictureFile()->save($file);
Integrity constraint violation: 1048 Column 'user_id' cannot be null
```
Hence why I thought `HasOne` simply doesn't support this type of relation and wasn't looking for `belongsTo` which does the right thing.

**Comment by jarektkaczyk** [comment]
@HolgerW1 You didn't get these relations. It is not going to work just by replacing the keys in the relation definition. It depends on the fields in your tables. I said, that it could be `hasOne`, but that means, the foreign key would be on the `files` table once again.

**Comment by GrahamCampbell** [comment]
Please continue this on the forums.
